src/engine.rs

```diff
@@ -148,9 +148,19 @@ where
                 let now = Instant::now();
                 let _t = StopWatch::new_with(&*ENGINE_WRITE_LEADER_DURATION_HISTOGRAM, now);
                 let file_context = self.pipe_log.fetch_active_file(LogQueue::Append);
+                // Flag on whether force to flush the current active file or not.
+                let mut force_rotate = false;
```

Comment:
Two things here, (1) I don't want the engine to care about "rotate" details. We can at most implement a `fn retriable` for the error which is true for disk full (2) We need to assume all buffered data is lost if `fsync` fails. To tolerate disk full we need a way to detect "if there's any buffered write from last write group". If so we must panic regardless of whether it's a disk full error or not.

I haven't thought of an elegant abstraction to handle both of these cases, maybe we can rethink it next week.

Reply:
About the second suggestion u mentioned above, I tried to figure out whether there exists an elegant solution to this corner case recently, but ended with failure. **I think it's hard to detect "if there's any buffered write from last write group".**

And I hold the view that if `fsync` is failed, there should be some hardware errs raised by IO device.
Refer to **https://wiki.postgresql.org/wiki/Fsync_Errors**, I think we also can take the same processing if `fsync` returns a unexpected error, **`panic`** is enough for it.

So, current `maybe_sync` is enought for `nospc` err:
- If `rotate` fails, it means:
  - There might have no free space or no stale files to reuse, should `panic`.
  - `fsync` is failed, as the previous annotations show, should `panic`.
- If `fsync` fails, just `panic`.

Reply:
The point is `fsync` can fail even if `pwrite` succeeds. But I agree in a practical sense this is unlikely to happen.